sparse_strips/vello_common/src/encode.rs

```diff
@@ -0,0 +1,819 @@
+                y_offset = -center.y as f32;
+                clamp_range = (start_angle, end_angle);
+
+                EncodedKind::Sweep(SweepKind)
+            }
+        };
+
+        let ranges = encode_stops(&stops, clamp_range.0, clamp_range.1, pad, opacity);
+
+        // This represents the transform that needs to be applied to the starting point of a
+        // command before starting with the rendering.
+        // First we need to account for a potential offset of the gradient (x_offset/y_offset), then
+        // we account for the fact that we sample in the center of a pixel and not in the corner by
+        // adding 0.5.
+        // Finally, we need to apply the _inverse_ transform to the point so that we can account
+        // for the transform on the gradient.
+        let transform = Affine::translate((x_offset as f64 + 0.5, y_offset as f64 + 0.5))
+            * self.transform.inverse();
+
+        // One possible approach of calculating the positions would be to apply the above
+        // transform to _each_ pixel that we render in the wide tile. However, a much better
+        // approach is to apply the transform once for the first pixel,
+        // and from then on only apply incremental updates to the current x/y position
+        // that we calculated in the beginning.
+        //
+        // Remember that we render wide tiles in column major order (i.e. we first calculate the
+        // values for a specific x for all Tile::HEIGHT y by incrementing y by 1, and then finally
+        // we increment the x position by 1 and start from the beginning). If we want to implement
+        // the above approach of incrementally updating the position, we need to calculate
+        // how the x/y unit vectors are affected by the transform, and then use this as the
+        // step delta for a step in the x/y direction.
+        let (x_advance, y_advance) = {
+            let scale_skew_transform = {
+                let c = transform.as_coeffs();
+                Affine::new([c[0], c[1], c[2], c[3], 0.0, 0.0])
+            };
+
+            let x_advance = scale_skew_transform * Point::new(1.0, 0.0);
+            let y_advance = scale_skew_transform * Point::new(0.0, 1.0);
+
+            (
+                Vec2::new(x_advance.x, x_advance.y),
+                Vec2::new(y_advance.x, y_advance.y),
+            )
+        };
+
+        let encoded = EncodedGradient {
+            kind,
+            transform,
+            x_advance,
+            y_advance,
+            ranges,
+            pad,
+            has_opacities,
+            clamp_range,
+        };
+
+        let idx = paints.len();
+        paints.push(encoded.into());
+
+        Paint::Indexed(IndexedPaint::new(idx))
+    }
+}
+
+/// Returns a fallback paint in case the gradient is invalid. The paint will be either
+/// black or contain the color of the first stop of the gradient.
+fn validate(gradient: &Gradient) -> Option<Paint> {
```

Comment:
```suggestion
fn validate(gradient: &Gradient) -> Result<(), Paint> {
```
would be a more idiomatic signature